File: test/ModularPipelines.UnitTests/Validation/ValidationInterfaceTests.cs
using ModularPipelines.Context;
using ModularPipelines.Validation;
using System.Reflection;

namespace ModularPipelines.UnitTests.Validation;

/// <summary>
/// Tests for validation interface structure.
/// </summary>
public class ValidationInterfaceTests
{
    public sealed class TestValidator : IPipelineValidator
    {
        public int Order => 0;

        public Task<ValidationResult> ValidateAsync(IServiceProvider services) =>
            Task.FromResult(ValidationResult.Success());
    }

    [Test]
    public async Task IPipelineValidationService_ShouldBeInternal()
    {
        var assembly = typeof(IModuleContext).Assembly;
        var iface = assembly.GetType("ModularPipelines.Validation.IPipelineValidationService");

        await Assert.That(iface).IsNotNull()
            .Because("IPipelineValidationService should exist");
        await Assert.That(iface!.IsPublic).IsFalse()
            .Because("IPipelineValidationService should be internal (implementation detail)");
    }

    [Test]
    public async Task IPipelineValidator_ShouldRemainPublic()
    {
        // IPipelineValidator is a user extension point - custom validators can implement it
        var validatorType = typeof(IPipelineValidator);

        await Assert.That(validatorType.IsPublic).IsTrue()
            .Because("IPipelineValidator should be public for custom implementations");
    }

    [Test]
    public async Task IPipelineValidator_ShouldHaveOrderAndValidateAsyncMembers()
    {
        var validatorType = typeof(IPipelineValidator);

        var orderProperty = validatorType.GetProperty("Order");
        await Assert.That(orderProperty).IsNotNull()
            .Because("IPipelineValidator should have Order property");

        var validateAsyncMethod = validatorType.GetMethod(nameof(IPipelineValidator.ValidateAsync));
        await Assert.That(validateAsyncMethod).IsNotNull()
            .Because("IPipelineValidator should have ValidateAsync method");
    }

    [Test]
    public async Task SpecializedValidatorInterfacesShouldBeInternal()
    {
        var assembly = typeof(IPipelineValidator).Assembly;
        var typeNames = new[]
        {
            "ModularPipelines.Validation.IDependencyValidator",
            "ModularPipelines.Validation.IModuleConfigurationValidator",
            "ModularPipelines.Validation.IOptionsValidator",
        };

        foreach (var typeName in typeNames)
        {
            var type = assembly.GetType(typeName);
            await Assert.That(type).IsNotNull();
            await Assert.That(type!.IsNotPublic).IsTrue();
        }
    }

    [Test]
    public async Task AddValidatorRegistersPublicValidatorExtensionPoint()
    {
        var builder = Pipeline.CreateBuilder();

        var result = builder.AddValidator<TestValidator>();
        var descriptor = builder.Services.Single(service => service.ServiceType == typeof(IPipelineValidator));

        await Assert.That(result).IsSameReferenceAs(builder);
        await Assert.That(descriptor.ImplementationType).IsEqualTo(typeof(TestValidator));
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Validation/ValidationInterfaceTests.fs
namespace ModularPipelines.UnitTests.FSharp.Validation

open System.Reflection
open ModularPipelines.Context
open ModularPipelines.Validation
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type ValidationInterfaceTests() =
    [<Test>]
    member _.IPipelineValidationService_ShouldBeInternal() = async {
        let assembly = typeof<IModuleContext>.Assembly
        let iface = assembly.GetType("ModularPipelines.Validation.IPipelineValidationService")

        do! check(Assert.That(iface).IsNotNull())
        do! check(Assert.That(iface.IsPublic).IsFalse())
    }

    [<Test>]
    member _.IPipelineValidator_ShouldRemainPublic() = async {
        let validatorType = typeof<IPipelineValidator>
        do! check(Assert.That(validatorType.IsPublic).IsTrue())
    }

    [<Test>]
    member _.IPipelineValidator_ShouldHaveOrderAndValidateMembers() = async {
        let validatorType = typeof<IPipelineValidator>

        let orderProperty = validatorType.GetProperty("Order")
        do! check(Assert.That(orderProperty).IsNotNull())

        let validateMethod = validatorType.GetMethod("Validate")
        do! check(Assert.That(validateMethod).IsNotNull())
    }
